src/Features/Core/Portable/ConvertAnonymousTypeToClass/AbstractConvertAnonymousTypeToClassCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,412 @@
+            // generate the same Equals/GetHashCode that our other IDE features generate.
+            var propertiesForEqualsAndGetHashCode = ImmutableArray<ISymbol>.CastUp(
+                properties.WhereAsArray(p => p.SetMethod == null));
+
+            var equalsAndGetHashCodeService = document.GetLanguageService<IGenerateEqualsAndGetHashCodeService>();
+                        
+            var equalsMethod = await equalsAndGetHashCodeService.GenerateEqualsMethodAsync(
+                document, namedTypeWithoutMembers, propertiesForEqualsAndGetHashCode, 
+                localNameOpt: ICodeDefinitionFactoryExtensions.OtherName, cancellationToken).ConfigureAwait(false);
+            var getHashCodeMethod = await equalsAndGetHashCodeService.GenerateGetHashCodeMethodAsync(
+                document, namedTypeWithoutMembers,
+                propertiesForEqualsAndGetHashCode, cancellationToken).ConfigureAwait(false);
+
+            var members = ArrayBuilder<ISymbol>.GetInstance();
+            members.AddRange(properties);
+            members.Add(constructor);
+            members.Add(equalsMethod);
+            members.Add(getHashCodeMethod);
+
+            var namedTypeSymbol = CreateNamedType(className, capturedTypeParameters, members.ToImmutableAndFree());
+            return namedTypeSymbol;
+        }
+
+        private static INamedTypeSymbol CreateNamedType(
+            string className, ImmutableArray<ITypeParameterSymbol> capturedTypeParameters, ImmutableArray<ISymbol> members)
+        {
+            return CodeGenerationSymbolFactory.CreateNamedTypeSymbol(
+                attributes: default, Accessibility.Internal, modifiers: default, 
+                TypeKind.Class, className, capturedTypeParameters, members: members);
+        }
+
+        private (ImmutableArray<IPropertySymbol>, ImmutableDictionary<IPropertySymbol, string>) GenerateProperties(INamedTypeSymbol anonymousType)
+        {
+            var originalProperties = anonymousType.GetMembers().OfType<IPropertySymbol>().ToImmutableArray();
+            var newProperties = originalProperties.SelectAsArray(GenerateProperty);
+
+            // If we changed the names of any properties, record that name mapping.  We'll
+            // use this to update reference to the old anonymous-type properties to the new
+            // names.
+            var builder = ImmutableDictionary.CreateBuilder<IPropertySymbol, string>();
+            for (var i = 0; i< originalProperties.Length; i++)
+            {
+                var originalProperty = originalProperties[i];
+                var newProperty = newProperties[i];
+
+                if (originalProperty.Name != newProperty.Name)
+                {
+                    builder[originalProperty] = newProperty.Name;
+                }
+            }
+
+            return (newProperties, builder.ToImmutable());
+        }
+
+        private IPropertySymbol GenerateProperty(IPropertySymbol prop)
```

Comment:
>GenerateProperty [](start = 32, length = 16)

static?

Reply:
Done (and several other methods as well).